src/Voron/Data/Tables/TableSchema.cs

```diff
@@ -328,31 +317,169 @@ public void Validate(FixedSizeSchemaIndexDef actual)
             }
         }
 
-        public TableSchema CompressValues(FixedSizeSchemaIndexDef etagSource, bool compress)
+        public class DynamicBTreeIndexDef : AbstractBTreeIndexDef
+        {
+            public override TableIndexType Type => TableIndexType.Dynamic;
+
+            public delegate ByteStringContext.Scope IndexValueAction(ByteStringContext context, ref TableValueReader value, out Slice slice);
+
+            public IndexValueAction IndexValueGenerator;
+
+            public override ByteStringContext.Scope GetSlice(ByteStringContext context, ref TableValueReader value,
+                out Slice slice)
+            {
+                return IndexValueGenerator(context, ref value, out slice);
+            }
+
+            public override ByteStringContext.Scope GetSlice(ByteStringContext context, TableValueBuilder value,
+                out Slice slice)
+            {
+                using (context.Allocate(value.Size, out var buffer))
+                {
+                    value.CopyTo(buffer.Ptr);
+                    var reader = value.CreateReader(buffer.Ptr);
+                    return IndexValueGenerator(context, ref reader, out slice);
+                }
+            }
+
+            public override byte[] Serialize()
+            {
+                // We serialize the Type enum as ulong to be "future-proof"
+                var castedType = (long)Type;
+
+                var serializer = new TableValueBuilder
+                {
+                    castedType,
+                    IsGlobal,
+                    Name
+                };
+
+                var methodNameBytes = Encodings.Utf8.GetBytes(IndexValueGenerator.Method.Name);
+                fixed (byte* ptr = methodNameBytes)
+                {
+                    serializer.Add(ptr, methodNameBytes.Length);
+                }
+
+                Debug.Assert(IndexValueGenerator.Method.DeclaringType?.AssemblyQualifiedName != null, 
+                    $"Invalid {nameof(IndexValueGenerator)} '{IndexValueGenerator.Method.Name}'");
+
+                var declaringTypeBytes = Encodings.Utf8.GetBytes(IndexValueGenerator.Method.DeclaringType.AssemblyQualifiedName);
```

Comment:
That is probably a bad idea, we want to skip the versioning here:

`System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089`

Because we want to be future proof.

Reply:
done